File: crates/daemon/src/services/cache.rs
//! The content-addressed download cache.

use proto::cache::{
    CacheClear, CacheEntry, CacheInfo, CacheInfoResult, CacheList, CacheListResult, CacheUsage,
};
use proto::Empty;

use crate::runtime::Channels;

pub(super) fn register(on: &mut Channels<'_>) {
    on.handle::<CacheInfo, _, _>(|_: Empty, ctx| async move {
        let cache = ctx.runtime.engine().cache();
        let usage = cache.usage();
        Ok(CacheInfoResult {
            path: cache.dir(),
            usage: CacheUsage {
                entries: usage.entries,
                bytes: usage.bytes,
            },
        })
    });

    on.handle::<CacheList, _, _>(|_: Empty, ctx| async move {
        let entries = ctx
            .runtime
            .engine()
            .cache()
            .entries()
            .into_iter()
            .map(|e| CacheEntry {
                checksum: e.checksum,
                size: e.size,
            })
            .collect();
        Ok(CacheListResult { entries })
    });

    on.handle::<CacheClear, _, _>(|_: Empty, ctx| async move {
        let freed = ctx.runtime.engine().cache().clear();
        tracing::info!(
            entries = freed.entries,
            bytes = freed.bytes,
            "cache cleared"
        );
        Ok(CacheUsage {
            entries: freed.entries,
            bytes: freed.bytes,
        })
    });
}

File: frontend/src/api/types/cache.ts
/** Mirrors `crates/proto/src/cache.rs`. */
import type { Checksum } from './download';

/** `Checksum` is flattened into the entry on the wire. */
export type CacheEntry = Checksum & {
  size: number;
};

export interface CacheUsage {
  entries: number;
  bytes: number;
}

/** `CacheUsage` is flattened beside `path` on the wire. */
export type CacheInfo = CacheUsage & {
  path: string;
};
